Bugfix/CM automatically imposed 21 day default loan length to loans without an end

pcustic opened the pull request:
## Description
Removing default loan period in the function that creates part of the OPDS feed that shows loan length for the book so it only uses `loan.end` (if it exists).

## Motivation and Context
For collections that don't restrict loan length it would happen that the loan in our database would be without an end but in the OPDS response we would see loan length of 21 days. Problem was that the CM would impose its 21 day default loan period (`STANDARD_DEFAULT_LOAN_PERIOD`). Client would see that the loan lasts for 21 days even though it's an indefinite loan.

https://www.notion.so/lyrasis/CM-automatically-imposes-default-21-day-loan-length-restriction-on-some-collection-types-a61e1e5775c245dca6042e6fab3c1a02

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
Added tests that checks `until` field of the OPDS feed and compares it to `loan.end`.
Also, manually borrowed book from BiblioBoard collection and checked that there is no until period in return data from the `.borrow` endpoint.

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have updated the documentation accordingly.
- [x] All new and existing tests passed.

tdilauro reviewed (commented):
This initial review probably reflects more than anything that I do not fully understand how we're populating `loans.end`.

@jonathangreen Do you have any insights into this off the top of your head?

pcustic commented on tests/api/test_opds.py:
```diff
@@ -1699,6 +1705,10 @@ def test_acquisition_links(self):
 
+        availability = opds_parser._xpath1(fulfill, "opds:availability")
+        assert _strftime(loan1.start) == availability.attrib.get("since")
+        assert loan1.end == availability.attrib.get("until")  # None
```
No, not really, I was thinking about it but didn't put it 🤷‍♂️  
I'll add it 👍

jonathangreen commented:
Off the top of my head, I don't know how `loan.end` is being calculated, but if I had to guess, its probably not being consistently used across the various integrations. I also share the concern expressed in @tdilauro review that this change may override the configured default loan period for collections that do allow it to be configured. 

I wonder if a better approach would be to either: 
- Allow the configuration of the default loan period, defaulting to unlimited. This seems like it may be the easiest to implement.
- Override the `default_loan_period` for collection types that don't have the configuration setting.

jonathangreen commented:
From a message thread in slack: 
> I have a sense that we don't want to enforce 21 day limit at all since it is only in the OPDS response and not in the actual DB that it is currently being enforced, but I am not sure if it is safe to just remove completely

I think the best way to approach this is to write some tests for the various cases with this limit as it is currently operating so we fully understand the current behaviour. For example, since its currently only sent in the OPDS feed, what happens after the default expiry if its shorter then the loan period for collections that allow this setting. Once we have a good set of tests documenting the current bahavior we can better reason about how we want to change it.

pcustic commented:
I reverted back the code to the current state and I added tests for loan duration in regards to collection configuration and `STANDARD_DEFAULT_LOAN_PERIOD`. It can be divided into 3 categories of loans from:
- collection without any loan duration configuration
- collection with loan duration configuration shorter than `STANDARD_DEFAULT_LOAN_PERIOD`
- collection with loan duration configuration longer than `STANDARD_DEFAULT_LOAN_PERIOD`

In each of those categories I added loans with different duration in regards to config and `STANDARD_DEFAULT_LOAN_PERIOD`.

When the `loan_end` is set (by distributer/collection) in the `LoanInfo` that value is reflected in the loan record in database and in `borrow` response inside `until` field of acquisition links. 

However, when the `loan_end` is not set the loan record in the database has no `loan.end` and we have two situations for the `until` field in the acquisition links:
a) it is set to `STANDARD_DEFAULT_LOAN_PERIOD` if there is no collection configuration
b) it is set to collection configuration loan duration if that is configured
I added `TODO`s in the code where we have discrepancy between `loan.end` and `until` fields - we can remove those when we decide what to do.

`loan_end` in `LoanInfo` is set per collection integration and it is not really standardised, in some places it is only data from integration, in some places it is combined with collection config and in some places it is just `None`.  

I am not sure if anything happens with loans after the `loan.end` or `until` period? I know I can still `fulfill` those loans and I didn't find something in the CM code that would depend on those two fields.

pcustic commented:
Also, currently tests added in previous commits (https://github.com/ThePalaceProject/circulation/pull/577/commits/4b4cc8cf0720c452c16dcb3de9d9cd40b293e098) fail because I reverted code to the original state.

pcustic commented:
@jonathangreen @tdilauro 
I tried to simplify the tests a bit so I parametrised them instead of having one huge test with all of the cases. It is still bunch of test cases so not sure will it still be a bit convoluted. Old version can be seen here.

As a refresher - we decided that I will set up some tests to see what is happening and then we will decide what to do. There isn't any actual fix for the problem in this PR. 
Loan duration in DB is set by collection (various implementations) and loan duration in OPDS response defaults to CM default when there is no loan duration set by collection so those instances are problematic.

jonathangreen reviewed (approved):
This looks good. 

Minor: but you probably don't need the changes to `poetry.lock` as part of this PR.
